lib/quicly.c

```diff
@@ -3873,29 +3873,50 @@ size_t quicly_send_retry(quicly_context_t *ctx, ptls_aead_context_t *token_encry
     return ret == 0 ? buf.off : SIZE_MAX;
 }
 
-static int send_handshake_flow(quicly_conn_t *conn, size_t epoch, quicly_send_context_t *s, int ack_only, int send_probe)
+static struct st_quicly_pn_space_t *setup_send_epoch(quicly_conn_t *conn, size_t epoch, quicly_send_context_t *s)
 {
     struct st_quicly_pn_space_t *ack_space = NULL;
-    int ret = 0;
 
     switch (epoch) {
     case QUICLY_EPOCH_INITIAL:
         if (conn->initial == NULL || (s->current.cipher = &conn->initial->cipher.egress)->aead == NULL)
-            return 0;
+            return NULL;
         s->current.first_byte = QUICLY_PACKET_TYPE_INITIAL;
         ack_space = &conn->initial->super;
         break;
     case QUICLY_EPOCH_HANDSHAKE:
         if (conn->handshake == NULL || (s->current.cipher = &conn->handshake->cipher.egress)->aead == NULL)
-            return 0;
+            return NULL;
         s->current.first_byte = QUICLY_PACKET_TYPE_HANDSHAKE;
         ack_space = &conn->handshake->super;
         break;
+    case QUICLY_EPOCH_0RTT:
+    case QUICLY_EPOCH_1RTT:
+        if (conn->application == NULL || conn->application->cipher.egress.key.header_protection == NULL)
+            return NULL;
+        if ((epoch == QUICLY_EPOCH_0RTT) == conn->application->one_rtt_writable)
+            return NULL;
+        s->current.cipher = &conn->application->cipher.egress.key;
+        s->current.first_byte = epoch == QUICLY_EPOCH_0RTT ? QUICLY_PACKET_TYPE_0RTT : QUICLY_QUIC_BIT;
+        ack_space = &conn->application->super;
+        break;
     default:
         assert(!"logic flaw");
-        return 0;
+        break;
     }
 
+    return ack_space;
+}
+
+static int send_handshake_flow(quicly_conn_t *conn, size_t epoch, quicly_send_context_t *s, int ack_only, int send_probe)
+{
+    struct st_quicly_pn_space_t *ack_space;
```

Comment:
send_space?

Reply:
Changed to `space`. We do not need to be verbose for local variables.

Reply:
SGTM. This variable is still `ack_space`.

Reply:
Ah, sorry. Fixed in 0cfef38.